### DISCUSSION TITLE
What is the appropriate pattern to split the store in modules?

### DISCUSSION THREAD

**Discussion by juliolmuller** [pmndrs/zustand#937] [Q&A]
First of all: congratulations to the core team for developing such an amazing tool. It was about time to have a tool to match Redux! :heart:

I understand that most frontend **store ** approaches/patterns relies on a "single source of truth". Therefore, the premise would be creating a single store and merge all the different parts (state and methods) in a single object when instantiating `zustand` store, like this:

```ts
import create from 'zustand';
import { IModuleA, stateModuleA, actionsModuleA } from 'moduleA';
import { IModuleB, stateModuleB, actionsModuleB } from 'moduleA';
import { IModuleC, stateModuleC, actionsModuleC } from 'moduleA';

export const useStore = create<IModuleA & IModuleB & IModuleC>((set, get) => ({
  ...stateModuleA,
  ...actionsModuleA(set, get),
   ...stateModuleB,
  ...actionsModuleB(set, get),
   ...stateModuleC,
  ...actionsModuleC(set, get),
}));
```

This pattern, however, exposes the state/methods to undesired collisions/overrides. But the most annoying thing is the "single-level" properties where the store is set on, so each state/action must be more descriptive like `{ createReport: () => {}, createFile^() => {}, createFoo: () => {}, ... }`. 

As a workaround to this, I wonder if the pattern of **creating multiple stores** is acceptable in `zustand` world, like this:

```ts
import create from 'zustand';
import { IModuleA, stateModuleA, actionsModuleA } from 'moduleA';
import { IModuleB, stateModuleB, actionsModuleB } from 'moduleA';
import { IModuleC, stateModuleC, actionsModuleC } from 'moduleA';

export const useA= create<IModuleA>((set, get) => ({
  ...stateModuleA,
  ...actionsModuleA(set, get),
}));

export const useB= create<IModuleB>((set, get) => ({
  ...stateModuleB,
  ...actionsModuleB(set, get),
}));

export const useC= create<IModuleC>((set, get) => ({
  ...stateModuleC,
  ...actionsModuleC(set, get),
}));
```

Wrapping up: what is the best pattern to create a multiple-modules-like store with this library?

**Comment by dai-shi** [comment]
First option is flux-like pattern: https://github.com/pmndrs/zustand/wiki/Flux-inspired-practice
Zustand is originally designed for this.

For second option, the community is working on a practice splitting into slices: https://github.com/pmndrs/zustand/wiki/Splitting-the-store-into-separate-slices
As you notice, `get` and `set` point to the entire store, not a slice.

Third option would be namespacing. This is an open field for third-party libraries. See: https://github.com/pmndrs/zustand/issues/178
[zustand-lens](https://github.com/dhmk083/dhmk-zustand-lens) seems promising.

**Reply by juliolmuller** [reply]
Amazing ideas!

So creating multiple stores is not an option? Would that be a bad practice? What are the risks of doing this and linking them with the imperative calls (`useA.fooAction(...)`)?

**Reply by dai-shi** [reply]
Oh, yeah. Creating multiple stores is an option too. The library doesn't block it but do anything about it, so stores are totally isolated. One of drawbacks is making a derived value from multiple stores is not easy. Invoking actions on other stores would not be an issue.

**Reply by rortan134** [reply]
Would there be any difference in performance (maybe memory-wise?) using the second method compared to the first one?

**Reply by dai-shi** [reply]
The first one and second one only differ in how to create a store. So, there's any difference in performance after creating it.
Performance-wise, a better solution could be multiple stores instead of a single store with slices. (It can achieve more scoped subscriptions.)

**Reply by teddybee** [reply]
Is there any solution to reach slices as separate objects in the store? I would like to reach them by groups in one store(single source of truth), and there can be a situation where some slices contains the same error field. I have tried like this, but it will not update the fields(at least I don`t see any changes in React components). 

```
export type StoreState = {product: ProductSlice , auth: AuthSlice}
export const useStore = create<StoreState>((...args) => ({
  product: ...createProductSlice(...args),
  auth: ...createAuthSlice(...args)
}));
```

**Reply by dai-shi** [reply]
Do you mean you want namespacing? #178

**Reply by dai-shi** [reply]
See also: #2647
